File: clients/go/conventions/goals/project_test.go
package goals_test

import (
	"encoding/json"
	"testing"

	"github.com/love-lena/sextant/clients/go/conventions/goals"
)

// TestProjectAppliesProofFilter is the proof that the Goals projection turns a
// stored goal into effective statuses in ONE place: an unproved stored "met"
// projects as in-progress and is NOT counted in the rollup; a proved "met"
// projects met and counts. This is the rule a UI must NOT reimplement.
func TestProjectAppliesProofFilter(t *testing.T) {
	arts := []goals.Artifact{
		{
			Name:     "goal.g1",
			Revision: 7,
			Record: json.RawMessage(`{"$type":"goal","northstar":"ship it","stream":"v0.5",` +
				`"criteria":[` +
				`{"id":"c1","text":"proved done","status":"met","owner":"sirius"},` + // proved below
				`{"id":"c2","text":"claimed done, no proof","status":"met"},` + // UNPROVED met
				`{"id":"c3","text":"in flight","status":"in-progress"}],` +
				`"review":{"state":"review"}}`),
		},
		// A proof artifact backing only c1.
		{Name: "the-proof", Revision: 1, Record: json.RawMessage(`{"title":"PR","relates":[{"goal":"g1","crit":"c1","kind":"proof"}]}`)},
		// A soft related ref to c2 — does NOT prove it.
		{Name: "a-note", Revision: 1, Record: json.RawMessage(`{"title":"note","relates":[{"goal":"g1","crit":"c2","kind":"related"}]}`)},
	}

	views := goals.Project(arts)
	if len(views) != 1 {
		t.Fatalf("Project returned %d views, want 1", len(views))
	}
	g := views[0]
	if g.ID != "g1" || g.Name != "goal.g1" || g.Northstar != "ship it" || g.Stream != "v0.5" {
		t.Errorf("view header wrong: %+v", g)
	}
	if g.Revision != 7 {
		t.Errorf("revision = %d, want 7", g.Revision)
	}
	if g.Review != "review" {
		t.Errorf("review = %q, want review", g.Review)
	}
	if len(g.Criteria) != 3 {
		t.Fatalf("criteria = %d, want 3", len(g.Criteria))
	}

	// c1: proved met → reads met, carries the proof evidence.
	if g.Criteria[0].Status != goals.StatusMet {
		t.Errorf("c1 effective status = %q, want met (proved)", g.Criteria[0].Status)
	}
	if len(g.Criteria[0].Evidence) != 1 || g.Criteria[0].Evidence[0].Kind != "proof" || g.Criteria[0].Evidence[0].Name != "the-proof" {
		t.Errorf("c1 evidence = %+v, want [the-proof proof]", g.Criteria[0].Evidence)
	}
	// c2: UNPROVED met → reads in-progress (the proof-filter downgrade), soft
	// related evidence present but not counted as proof.
	if g.Criteria[1].Status != goals.StatusInProgress {
		t.Errorf("c2 effective status = %q, want in-progress (unproved met downgraded)", g.Criteria[1].Status)
	}
	if len(g.Criteria[1].Evidence) != 1 || g.Criteria[1].Evidence[0].Kind != "related" {
		t.Errorf("c2 evidence = %+v, want [a-note related]", g.Criteria[1].Evidence)
	}

	// Rollup: only c1 counts as met (1 of 3), derived after the filter.
	if g.Rollup.Met != 1 || g.Rollup.Total != 3 {
		t.Errorf("rollup = %+v, want Met=1 Total=3 (the unproved met does NOT count)", g.Rollup)
	}
}

func TestProjectIgnoresNonGoals(t *testing.T) {
	arts := []goals.Artifact{
		{Name: "a-doc", Record: json.RawMessage(`{"title":"not a goal","body":"x"}`)},
		{Name: "goal.g2", Record: json.RawMessage(`{"northstar":"y","criteria":[]}`)},
	}
	views := goals.Project(arts)
	if len(views) != 1 || views[0].ID != "g2" {
		t.Fatalf("Project should surface only goal.g2; got %+v", views)
	}
}

File: conventions/goal/ts/test/project.test.ts
// Unit tests for the Goals PROJECTION — the co-equal peer of Go's
// project_test.go (ADR-0044). The SAME fixtures and the SAME proof-filter
// assertions, so the TS project() turns a stored goal into effective statuses
// IDENTICALLY to the Go goals.Project the dash backend used to serve. This is the
// rule a UI must NOT reimplement; the dash now calls project() directly in the
// browser, so these pin that the rule did not drift in the port.

import { test } from "node:test";
import assert from "node:assert/strict";
import type { JSONValue } from "@sextant/sdk";
import { project, StatusMet, StatusInProgress, type Artifact } from "../src/index.js";

// rec parses a JSON string into the JSONValue a record crosses as (the SDK's
// natural currency), matching how the dash hands project() artifact records.
function rec(s: string): JSONValue {
  return JSON.parse(s) as JSONValue;
}

test("project applies the proof-filter: an unproved met downgrades and is not counted", () => {
  const arts: Artifact[] = [
    {
      name: "goal.g1",
      revision: 7,
      record: rec(
        `{"$type":"goal","northstar":"ship it","stream":"v0.5",` +
          `"criteria":[` +
          `{"id":"c1","text":"proved done","status":"met","owner":"sirius"},` + // proved below
          `{"id":"c2","text":"claimed done, no proof","status":"met"},` + // UNPROVED met
          `{"id":"c3","text":"in flight","status":"in-progress"}],` +
          `"review":{"state":"review"}}`,
      ),
    },
    // A proof artifact backing only c1.
    { name: "the-proof", revision: 1, record: rec(`{"title":"PR","relates":[{"goal":"g1","crit":"c1","kind":"proof"}]}`) },
    // A soft related ref to c2 — does NOT prove it.
    { name: "a-note", revision: 1, record: rec(`{"title":"note","relates":[{"goal":"g1","crit":"c2","kind":"related"}]}`) },
  ];

  const views = project(arts);
  assert.equal(views.length, 1);
  const g = views[0]!;
  assert.equal(g.id, "g1");
  assert.equal(g.name, "goal.g1");
  assert.equal(g.northstar, "ship it");
  assert.equal(g.stream, "v0.5");
  assert.equal(g.revision, 7);
  assert.equal(g.review, "review");
  assert.equal(g.criteria.length, 3);

  // c1: proved met → reads met, carries the proof evidence.
  assert.equal(g.criteria[0]!.status, StatusMet);
  assert.equal(g.criteria[0]!.evidence?.length, 1);
  assert.equal(g.criteria[0]!.evidence?.[0]!.kind, "proof");
  assert.equal(g.criteria[0]!.evidence?.[0]!.name, "the-proof");

  // c2: UNPROVED met → reads in-progress (the proof-filter downgrade); the soft
  // related evidence is present but not counted as proof.
  assert.equal(g.criteria[1]!.status, StatusInProgress);
  assert.equal(g.criteria[1]!.evidence?.length, 1);
  assert.equal(g.criteria[1]!.evidence?.[0]!.kind, "related");

  // Rollup: only c1 counts as met (1 of 3), derived after the filter.
  assert.equal(g.rollup.met, 1);
  assert.equal(g.rollup.total, 3);
});

test("project ignores non-goal artifacts (they may still be proof sources)", () => {
  const arts: Artifact[] = [
    { name: "a-doc", revision: 1, record: rec(`{"title":"not a goal","body":"x"}`) },
    { name: "goal.g2", revision: 1, record: rec(`{"northstar":"y","criteria":[]}`) },
  ];
  const views = project(arts);
  assert.equal(views.length, 1);
  assert.equal(views[0]!.id, "g2");
});

test("project sorts views by artifact name and wires goal-level evidence", () => {
  const arts: Artifact[] = [
    { name: "goal.zeta", revision: 1, record: rec(`{"northstar":"z","criteria":[{"id":"c1","text":"t","status":"not-started"}]}`) },
    { name: "goal.alpha", revision: 1, record: rec(`{"northstar":"a","criteria":[{"id":"c1","text":"t","status":"not-started"}]}`) },
    // a goal-level relation (a goal, no crit) → goal-level evidence on alpha.
    { name: "design-doc", revision: 1, record: rec(`{"title":"design","relates":[{"goal":"alpha","kind":"related"}]}`) },
  ];
  const views = project(arts);
  assert.deepEqual(views.map((v) => v.name), ["goal.alpha", "goal.zeta"]);
  const alpha = views[0]!;
  assert.equal(alpha.evidence?.length, 1);
  assert.equal(alpha.evidence?.[0]!.name, "design-doc");
});
